File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def verify(node:Optional[TreeNode], left_bound:Optional[int], right_bound: Optional[int]) -> bool:
    if node == None:
        return True
    val = node.val
    
    if left_bound !=None:
        if left_bound >= val:
            return False
        
    if right_bound != None:
        if right_bound <= val:
            return False
    
    left = verify(node.left, left_bound, val) 
    # print("l:{}, res:{}".format(node.left, left))
    right = verify(node.right, val, right_bound)
    # print("r:{}, res:{}".format(node.right, right))
        
    return left and right
    
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return verify(root, None, None)

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl TreeNode{
    pub fn verify(&self, left_bound: Option<i32>, right_bound:Option<i32>) -> bool{
        let val = self.val;
        
        if let Some(l) = left_bound{
            if val <= l{
                return false;
            }
        }
        
        if let Some(r) = right_bound{
            if val >=r {
                return false;
            }
        }
        
        let mut res = true;
        
        if let Some(ref left) = &self.left{
            let is_valid = left.borrow().verify(left_bound, Some(val));
            
            res = res && is_valid;
        }
        
        if let Some(ref right) = &self.right{
            let is_valid = right.borrow().verify(Some(val), right_bound);
            
            res = res && is_valid;
        }
        
        res
    }
}

impl Solution {
    pub fn is_valid_bst(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        if let Some(node) = root{
            node.borrow().verify(None, None)
        }else{
            true
        }
    }
}